integration-cli/docker_cli_swarm_test.go

```diff
@@ -946,10 +951,164 @@ func (s *DockerSwarmSuite) TestSwarmLeaveLocked(c *check.C) {
+
+	// the ones that got the update are now set to unlocked
+	for _, d := range []*SwarmDaemon{d1, d3} {
+		c.Assert(d.Restart(), checker.IsNil)
+		checkLockStatus(c, d, false)
+	}
+
+	// d2 still locked
+	checkLockStatus(c, d2, true)
+
+	// unlock it and give a sec to catch up
+	cmd := d2.command("swarm", "unlock")
+	cmd.Stdin = bytes.NewBufferString(unlockKey)
+	out, err := cmd.CombinedOutput()
+	c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
+	checkLockStatus(c, d2, false)
+
+	// d2 is set to not be locked
+	c.Assert(d2.Restart(), checker.IsNil)
+	checkLockStatus(c, d2, false)
+}
+
+func (s *DockerSwarmSuite) TestSwarmJoinPromoteLocked(c *check.C) {
+	d1 := s.AddDaemon(c, true, true)
+
+	// enable autolock
+	outs, err := d1.Cmd("swarm", "update", "--autolock")
+	c.Assert(err, checker.IsNil, check.Commentf("out: %v", outs))
+
+	c.Assert(outs, checker.Contains, "docker swarm unlock")
+
+	var unlockKey string
+	for _, line := range strings.Split(outs, "\n") {
+		if strings.Contains(line, "SWMKEY") {
+			unlockKey = strings.TrimSpace(line)
+			break
+		}
+	}
+
+	c.Assert(unlockKey, checker.Not(checker.Equals), "")
+
+	outs, err = d1.Cmd("swarm", "unlock-key", "-q")
+	c.Assert(outs, checker.Equals, unlockKey+"\n")
+
+	// joined workers start off unlocked
+	d2 := s.AddDaemon(c, true, false)
+	c.Assert(d2.Restart(), checker.IsNil)
+	checkLockStatus(c, d2, false)
+
+	// promote worker
+	outs, err = d1.Cmd("node", "promote", d2.Info.NodeID)
+	c.Assert(err, checker.IsNil)
+	c.Assert(outs, checker.Contains, "promoted to a manager in the swarm")
+
+	// join new manager node
+	d3 := s.AddDaemon(c, true, true)
+
+	// wait for both of them to catch up
+	time.Sleep(1 * time.Second)
+
+	// all nodes are locked
+	for _, d := range []*SwarmDaemon{d1, d2, d3} {
+		c.Assert(d.Restart(), checker.IsNil)
+		checkLockStatus(c, d, true)
+
+		cmd := d.command("swarm", "unlock")
+		cmd.Stdin = bytes.NewBufferString(unlockKey)
+		out, err := cmd.CombinedOutput()
+		c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
+		checkLockStatus(c, d, false)
+	}
+
+	// demote manager back to worker - workers are not locked
+	outs, err = d1.Cmd("node", "demote", d3.Info.NodeID)
+	c.Assert(err, checker.IsNil)
+	c.Assert(outs, checker.Contains, "demoted in the swarm")
+
+	// verify that it's been demoted
+	out, err := d1.Cmd("node", "ls", "--filter", "id="+d3.Info.NodeID)
+	c.Assert(err, checker.IsNil)
+	lines := strings.Split(strings.TrimSpace(out), "\n")
+	c.Assert(len(lines), checker.GreaterThan, 0)
+	columns := strings.Fields(lines[len(lines)-1])
+	c.Assert(columns, checker.HasLen, 4) // if it was a manager it'd have a manager status field
+
+	c.Assert(d3.Restart(), checker.IsNil)
+	checkLockStatus(c, d3, false)
+}
+
 func (s *DockerSwarmSuite) TestSwarmRotateUnlockKey(c *check.C) {
-	d := s.AddDaemon(c, true, true)
+	d1 := s.AddDaemon(c, true, true) // leader - don't restart this one, we don't want leader election delays
```

Comment:
On second thought, I'd rather add this as a separate test in addition to the existing `TestSwarmRotateUnlockKey`. Not restarting `d1` does change what we're testing a bit. The original test was really good at finding problems.